cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/CoreSchedulerService.java

```diff
@@ -111,21 +121,77 @@ public void addSchedules(final Iterable<? extends ProgramSchedule> schedules) th
       @Override
       public Void run(ProgramScheduleStoreDataset store) throws AlreadyExistsException {
         store.addSchedules(schedules);
+        for (ProgramSchedule schedule : schedules) {
+          try {
+            scheduler.addProgramSchedule(schedule);
+          } catch (SchedulerException e) {
+            // TODO: temporarily catch the SchedulerException and throw AlreadyExistsException.
```

Comment:
If the schedule is successfully added to Quartz, then no exception will be thrown here. If an exception is thrown here, that means the schedule is not successfully added and it does not exist in Quartz, so there won't be `AlreadyExistException the` next time when adding this schedule to Quartz. For both Quartz and StreamSizeScheduler, they don't check whether a schedule already exists when they try to persist, so it's also safe to try adding a schedule again. If we try to delete the schedule in the catch block, there will be `NotFoundException` since it's not present in Quartz.

 Also, I think it's ok for Quartz and StreamSizeScheduler to persist a schedule but fail to add them to the scheduler and causes ProgramScheduleStore to rollback. If the scheduler restarts and read the schedules in `DatasetBasedStreamSizeScheduleStore` or `DatasetBasedTimeScheduleStore`, but it is not present in the ProgramScheduleStore, then on the subscriber side, nothing will happen.